libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/idp/IDPManager.kt

```diff
@@ -0,0 +1,267 @@
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission of salesforce.com, inc.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.auth.idp
+
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import android.webkit.WebView
+import com.salesforce.androidsdk.accounts.UserAccount
+import com.salesforce.androidsdk.app.SalesforceSDKManager
+import com.salesforce.androidsdk.auth.idp.IDPSPMessage.*
+import com.salesforce.androidsdk.auth.idp.interfaces.IDPManager.Status
+import com.salesforce.androidsdk.auth.idp.interfaces.IDPManager.StatusUpdateCallback
+import com.salesforce.androidsdk.util.LogUtil
+import com.salesforce.androidsdk.util.SalesforceSDKLogger
+
+
+/**
+ * Class to capture state related to IDP initiated login flow
+ * e.g. the context (activity) that started it
+ */
+internal class IDPInitiatedLoginFlow private constructor(context:Context, val user:UserAccount, val spConfig: SPConfig, val onStatusUpdate:(Status) -> Unit) : ActiveFlow(context) {
+    companion object {
+        val TAG = IDPInitiatedLoginFlow::class.java.simpleName
+        fun kickOff(idpManager:IDPManager, context: Context, user: UserAccount, spConfig: SPConfig, onStatusUpdate: (Status) -> Unit) : IDPInitiatedLoginFlow {
+            SalesforceSDKLogger.d(TAG, "Kicking off idp initiated login flow from ${context} for user:${user}, sp app:${spConfig.appPackageName}")
+
+            val activeFlow = IDPInitiatedLoginFlow(context, user, spConfig, onStatusUpdate)
+            val idpLoginRequest = IDPLoginRequest(orgId = user.orgId, userId = user.userId).also {
+                activeFlow.addMessage(it)
+            }
+            idpManager.send(context, idpLoginRequest, spConfig.appPackageName)
+            onStatusUpdate(Status.LOGIN_REQUEST_SENT_TO_SP)
+            return activeFlow
+        }
+    }
+
+}
+
+/**
+ * Class handling IDP operations within an IDP app
+ */
+internal class IDPManager(
+    val allowedSPApps: List<SPConfig>,
+    // the following allows us to decouple IDPManager from other part of the SDK and make it easier to test
+    val sdkMgr: SDKManager,
```

Comment:
To make testing possible IDPManager does not directly call SalesforceSDKManager or other part of the SDK. Even sendBroadcast is passed here.
There is however a secondary constructor (line 94) that wires all those things in. And that's the one we call from SalesforceSDKManager when creating the idpManager instance.